examples/service_worker/src/lib.rs

```diff
@@ -134,213 +86,232 @@ struct PushSubscriptionKeys {
-    FailedToRegisterServiceWorker(ServiceWorkerError),
-    FailedToUnregisterServiceWorker(ServiceWorkerError),
+    WorkerActivated(web_sys::ServiceWorkerRegistration),
+    SubscriptionReceived(Option<PushSubscription>),
+    WorkerRegistrationFailed(ServiceWorkerError),
+    UnregisterWorker,
+    WorkerUnregistered(Option<ServiceWorkerError>),
     RequestNotificationPermission,
-    ServiceWorkerActivated,
-    SubscriptionRetrieved(PushSubscription),
+    NotificationPermissionRequested(Result<web_sys::NotificationPermission, ServiceWorkerError>),
+    Subscribed(Result<PushSubscription, ServiceWorkerError>),
     SendMessage,
-    SetServiceWorker(web_sys::ServiceWorkerRegistration),
-    StatusUpdate(ServiceWorkerStatus),
-    UnregisterServiceWorker,
-    UnregisteredServiceWorker,
+    ClearCache,
+    CacheCleared(Option<ServiceWorkerError>),
 }
 
+#[derive(Debug)]
+pub enum ServiceWorkerError {
+    DeleteCache(JsValue),
+    InvalidPermissions,
+    WorkerRegistration(JsValue),
+    RequestPermission(JsValue),
+    SerdeJson(serde_json::error::Error),
+    WorkerUnregistration(Option<JsValue>),
+}
+
+#[allow(clippy::too_many_lines, clippy::match_same_arms)]
 fn update(msg: Msg, model: &mut Model, orders: &mut impl Orders<Msg>) {
     match msg {
-        Msg::CacheCleared => {
-            model.message = Some("Cache cleared.".into());
+        Msg::WorkerActivated(worker_registration) => {
+            log!("Worker activated.");
+            model.message = Some("Service worker activated!".to_owned());
+
+            let push_manager = worker_registration
+                .push_manager()
+                .expect("get `PushManager`");
+
+            if let Ok(get_subscription) = push_manager.get_subscription().map(JsFuture::from) {
+                orders.perform_cmd(async move {
+                    Msg::SubscriptionReceived({
+                        get_subscription
+                            .await
+                            .ok()
+                            .and_then(|subscription| subscription.into_serde().ok())
+                    })
+                });
+            };
+
+            model.worker_data = Some(WorkerData {
+                notifications_granted: web_sys::NotificationPermission::Granted
+                    == web_sys::Notification::permission(),
+                push_subscription: None,
+                registration: worker_registration,
+            });
         }
-        Msg::FailedToClearCache(e) => {
-            model.error = Some(format!("{:?}", e));
+        Msg::SubscriptionReceived(subscription) => {
+            let worker_data = match &mut model.worker_data {
+                None => return,
+                Some(worker_data) => worker_data,
+            };
+            worker_data.push_subscription = subscription;
         }
-        Msg::ClearCache => {
+        Msg::WorkerRegistrationFailed(worker_error) => {
+            model.error = Some(log_error(&worker_error));
```

Comment:
This seems to be a recurring pattern. It's too bad we can't use the `?` operator.

Reply:
I don't know how we should refactor it to make it simpler and still readable/flexible.
You can try something once this PR is merged, however I think it should be good enough.